util/annotations/annotations.go

```diff
@@ -118,12 +149,20 @@ type annoErr struct {
 	PositionRange posrange.PositionRange
 	Err           error
 	Query         string
+	Min           []float64
```

Comment:
We've discussed this IRL, I think it will be tricky to come up with a good generic data structure that handles all future use cases for annotations. What you need is an interface that has  `Merge` function. The implementation would do a type assertion and return error on failure and do the merge when the type assert is successful. And `Annotations` `Add` and `Merge` could use the interface. We wouldn't expect error from merge, since the `key` corresponds to the actual type I guess.

Come to think of it, you could also add a `Level` function to the interface that could tell you if the annotation is an `Info` or `Warning` .

Reply:
The info and warning distinction can be made with `errors.Is` (as we modeled annotations as Go errors). (#14327 shows it in action, but you see concerns in the review that it might be too tricksy.)

However, given that all annotations are just treated as Go errors, we can change the implementation for just some of them. Maybe it would be cleaner to let `NewHistogramQuantileForcedMonotonicityInfo` return an implementation specific to that type of annotation and don't even try to be generic.